runtime/common/src/xcm_impl.rs

```diff
@@ -235,6 +236,34 @@ impl<FixedRate: Get<u128>, R: TakeRevenue, M: BuyWeightRate> Drop for FixedRateO
 	}
 }
 
+pub struct XcmExecutor<Config: xcm_executor::Config, AccountId, Balance, AccountIdConvert, EVMBridge>(
+	PhantomData<(Config, AccountId, Balance, AccountIdConvert, EVMBridge)>,
+);
+
+impl<
+		Config: xcm_executor::Config,
+		AccountId: Clone,
+		Balance,
+		AccountIdConvert: xcm_executor::traits::Convert<MultiLocation, AccountId>,
+		EVMBridge: module_support::EVMBridge<AccountId, Balance>,
+	> ExecuteXcm<Config::RuntimeCall> for XcmExecutor<Config, AccountId, Balance, AccountIdConvert, EVMBridge>
+{
+	fn execute_xcm_in_credit(
+		origin: impl Into<MultiLocation>,
+		message: Xcm<Config::RuntimeCall>,
+		weight_limit: XcmWeight,
+		weight_credit: XcmWeight,
+	) -> Outcome {
+		let origin = origin.into();
+		let account = AccountIdConvert::convert(origin.clone());
+		if let Ok(account) = account {
+			EVMBridge::push_origin(account);
+			defer!(EVMBridge::pop_origin());
```

Comment:
I don't feel like `ExtrinsicOrigin ` needs to be an array

Reply:
yeah it doesn't have to. we can just add a clear_origin method

however I am slightly bit worried about reentrancy issue. e.g. if someone send a tx somehow trigger XcmExecution, the evm origin will be cleared and follow execution will not have the correct origin.

Reply:
I think this defer may execute when the `if` out of scope so what wouldn't work...

Reply:
i might miss something but here it seems nothing happened between pushing and popping the origin

Reply:
pop is defered using `defer`macro

Reply:
I think it's only defered in the `if` scope and will be executed before xcm executor

Reply:
https://github.com/paritytech/substrate/blob/master/primitives/core/src/defer.rs#L65-L73
Yes. `defer` executes in the `if` scope